### PR TITLE
Fix & enable irept pretty-printer

### PR DESCRIPTION
**Pull request by JohnDumbell**
Makes minor changes to the existing irept pretty-printer to make it work in CLion and then enable it. Marked as do-not-merge for now as I start using it locally to make sure there's no obvious problems.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by thk123** [comment]
AFAICT this doesn't print the `.id()` for ireps that have children - I think this is a bit of a problem - can we at least have it is as a fake entry in the children:

```
id symbol
[0] ..
[identifier] ...
```
?

Will definitely give this a play though!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6e96ad595e097f6c52bc60d209721dcb2b0aacba).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123400062

**Comment by JohnDumbell** [comment]
I'm going to make it so that it goes child: {id}, so:

`0: symbol`
`statement: assignment`

Means we can drop the fake child element, condense the tree a little and also add the missing info back in. Will make things much clearer, even if we are abusing the heck out of the pretty-printer at this stage.

File: scripts/pretty-printers/gdb/pretty_printers.py [hunk 1]
```diff
@@ -125,8 +141,14 @@ def children(self):
                     if node.dereference()["_M_right"] != parent:
                         node = parent
 
+        # If we have no children it means we're actually a value node, so
+        # return the value of the node and a throw-away second value.
+        if sub_count == 0 and named_sub_count == 0:
+            yield (self.to_string(), self.val)
```
Comment (NathanJPhillips) [inline scripts/pretty-printers/gdb/pretty_printers.py]: Why not just not return any children?

**Review by NathanJPhillips — COMMENTED** [review]
Looks good but can we see the output in plain old GDB before approving?

**Comment by JohnDumbell** [comment]
Pushed a change that means we get a more succinct tree in CLion, like so:

cleaned_pretty_printer

(Nodes with `[...]` can be expanded, quoted things are values)

But as the pretty-printer is an array it makes the `p` command in GDB directly a little less useful, as it omits any data about the position which we use to add more context: 

`{<exprt> = {<irept> = "code" = {"nil", "symbol" = {"code" = {"empty", "", "1"}, "__CPROVER_initialize"}, "arguments", "empty", "nil", "function_call"}, <No data fields>}, <No data fields>}`

I'll try the map quickly and see if that can work, but I really like the way it looks right now.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 350ec400c21ad65e93ec7bc5cec549fe21ec2a14).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123501029

**Comment by JohnDumbell** [comment]
I've added the option to switch between GDB/CLion pretty-printing mode as I didn't want to sacrifice either view. Example of the CLion mode is above, this is what GDB mode looks like:

`$1 = {<exprt> = {<irept> = "code" = {[0] = "nil", [1] = "symbol" = {[type] = "code" = {[return_type] = "empty", [parameters] = "", [#java_method_type] = "1"}, [identifier] = "__CPROVER_initialize"}, [2] = "arguments", [type] = "empty", [#source_location] = "nil", [statement] = "function_call"}, <No data fields>}, <No data fields>}`

File: scripts/pretty-printers/gdb/pretty_printers.py (hunk 1 above)
Comment (JohnDumbell) [reply scripts/pretty-printers/gdb/pretty_printers.py]: It needed this originally, doesn't anymore so has been removed.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c749cd4435c63280d51fa939938b9ab29b8c8825).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123508581

File: scripts/pretty-printers/gdb/options.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+  "clion_pretty_printers": true
+}
```
Comment (owen-mc-diffblue) [inline scripts/pretty-printers/gdb/options.json:3]: Missing newline :smile:

**Comment by JohnDumbell** [comment]
@thk123 Can I get an approval please, want to merge this soon. Owens been helping me test, and I've been able to use it personally for a few days and seems solid enough for general use.

Three things that are still on the to-do at this point:
1. Fix `instructiont` properly. This might take a while.
2. Make it so you don't have to expand a hierarchy to get the useful info. Shouldn't be too tricky, but it's a minor annoyance right now.
3. Add an ID to the root node that mimics how it works otherwise (`[symbol]`, etc), as it's the only piece of data that's not currently shown. (Adding a fake ID node to the top node is clumsy and fiddly, adding one to every node is just bloated, all other options require some experimentation which right now I don't have time for.

**Review by owen-mc-diffblue — APPROVED** [review]

**Review by thk123 — APPROVED** [review]
Truly you are doing the Lords work. I just gave this a spin and it is amazing :heart_eyes: 

image

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bfd2313ce9de869eddf7ef9822d20881861575f4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128316637

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: c6f2844d487a2dda704f5fa69e8b014d2c909ccd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128327640
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3a344a46883f075fdebce890ba22cdc0e3b8d1ae).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128331097

**Pull request merged**